clang/lib/CIR/CodeGen/CIRGenExprCXX.cpp

```diff
@@ -1048,7 +1048,10 @@ void CIRGenFunction::emitNewArrayInitializer(
     QualType AllocType = E->getAllocatedType();
     if (const ConstantArrayType *CAT = dyn_cast_or_null<ConstantArrayType>(
             AllocType->getAsArrayTypeUnsafe())) {
-      llvm_unreachable("NYI");
+      ElementTy = convertTypeForMem(AllocType);
+      auto CastOp = builder.createPtrBitcast(CurPtr.getPointer(), ElementTy);
+      CurPtr = Address(CastOp, ElementTy, CurPtr.getAlignment());
```

Comment:
The traditional clang codegen had `CurPtr = CurPtr.withElementType(ElementTy);` here, but that failed CIR verification because when we got an mlir::Value from CurPtr it didn't reflect the type change. I'll post a separate PR to hide the bitcast inside Address::withElementType() as discussed on Discord.